Learn/Extensions/NSNumber.swift

```diff
@@ -0,0 +1,26 @@
+//
+//  NSNumber.swift
+//  Learn
+//
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+
+extension NSNumber: Comparable {
+    public static func < (lhs: NSNumber, rhs: NSNumber) -> Bool {
+        return lhs.compare(rhs) == .orderedAscending
+    }
+}
+
+
+extension NSNumber {
+    convenience init?(value: Double?) {
+        if let value = value {
+            self.init(value: value)
+        } else {
+            return nil
+        }
+    }
```

Comment:
This is not overloading NSNumber.init, but rather providing another initializer for optionals, one that returns nil if a nil was passed in.  That seems straightforward and obvious to me at least.  Using map for a singular value, on the other hand, feels contorted.

Reply:
The same argument could be made for any initializer, or more generally, any function: "apply this operation on the operands, but only if they're non-nil; if they're nil, just return nil".
More abstractly, this describes lifting a function of type `(A) -> B` into a function of type `(A?) -> B?` that "fails" if its argument is nil. That's exactly the type signature and purpose of `map` on Optional, and in fact very closely related to Swift's optional chaining operator.

If you feel strongly about this, the initializer is fine with me—but the definition of `map` on Optional is no contorsion. It's exactly the same level of abstraction as `map` on Array.
Happy to go into more detail here if it helps; this is a fascinating subject rooted in intersection of category theory and functional programming.

Reply:
You kids get off my lawn!  

Ok, I did some reading on map & flatMap on optionals. I tend to prefer code that reads more clearly, rather than "clever" code. I'd have less of an issue with this if the practice of using map on optionals was more widespread.  Knowing how it works, I'm less opposed to it, but it does still feel like a bit of an obscurity.

Reply:
And, after thinking about this more, I do like that it solves a not uncommon problem in a generic way.  The name is the thing that most throws me off.  Something like value.optionallyApply(myfunc) seems like it would be more obvious.  

Additionally, seeing a lot of this kind of thing might be a code smell, in that our code generally shouldn't have a lot of optional values lying around.

Reply:
I agree the name is less intuitive for Optional and hinders discoverability.

The benefit with using the term of art across types is that the generic abstraction becomes apparent:
`map` on Optional: `(A?, (A) -> B) -> B?`
`map` on Array: &nbsp; &nbsp; &nbsp;`([A], (A) -> B) -> [B]` 
`map` generically:  &nbsp;`(F<A>, (A) -> B) -> F<B>`

The overrides PR defines [`map` on DailyValueSchedule](https://github.com/tidepool-org/LoopKit/blob/c8e7dde4b659a3eb81ba5dc2e360621895e62f19/LoopKit/DailyValueSchedule.swift#L169):
`(DailyValueSchedule<A>, (A) -> B) -> DailyValueSchedule<B>`

The shared name allows someone familiar with concept of `map` in other contexts to see `map` on a new type and know what the function does without additional documentation. `flatMap` describes a related pattern where one layer of the structure is "flattened" when applying the transformation.

With Optional in particular, these pattern crop up often because UIKit classes often vend properties with Optional types for unclear reasons, like UITextField's `text: String?`, which doesn't have an obvious semantic distinction between `nil` and `""`. The optionality of these values then percolates down when applying transformations, which is unfortunate.